### PR TITLE
feat: First working version

### PR DESCRIPTION
**Pull request by damaru-inc**
This is the first commit that contains code that is ready for an early review.

It is not production-ready, but it works with simple documents.

### CODE REVIEW

File: hooks/post-process.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// vim: set ts=2 sw=2 sts=2 expandtab :
+const fs = require('fs');
+const path = require('path');
+const _ = require('lodash');
+
+module.exports = register => {
+  register('generate:after', generator => {
+    const asyncapi = generator.asyncapi;
+    const info = asyncapi.info();
+    //let package = generator.templateParams['javaPackage'];
```
Comment (fmvilas) [inline hooks/post-process.js:10]: 🚜

File: hooks/post-process.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    let extensions = info.extensions();
+
+    for (schema in asyncapi.components().schemas()) {
+      //console.log("schema: " + schema);
```
Comment (fmvilas) [inline hooks/post-process.js:14]: 🚜

File: lib/templateUtil.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
+// This contains functions taht are common to both the all.js filter and the post-process.js hook.
+var _ = require('lodash');
+
+class TemplateUtil {
+
+	// This returns a valid Java class name.
```
Comment (fmvilas) [inline lib/templateUtil.js:6]:
```suggestion
	// This returns a valid Python class name.
```

File: lib/templateUtil.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	getClassName(name) {
+		let ret = this.getIdentifierName(name);
+		return _.upperFirst(ret);
+	}
+
+	// This returns a valid Java identifier name.
```
Comment (fmvilas) [inline lib/templateUtil.js:12]:
```suggestion
	// This returns a valid Python identifier name.
```

File: lib/templateUtil.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	/*
+	By default, the 'view' is 'client', which means that when the doc says subscribe, we publish.
+	By setting the view to 'provider', when the doc says subscribe, we subscribe.
+	*/
+	isProvidererView(info, params) {
+    let view = this.getParamOrDefault(info, params, 'view', 'x-view', 'client');
+    //console.log('isProviderView:');
+    //console.log(view);
```
Comment (fmvilas) [inline lib/templateUtil.js:58]: 🚜

File: lib/templateUtil.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	/*
+	See isProviderView above.
+	This returns true if the operation should physically subscribe, based on the 'view' param.
+	*/
+	isRealSubscriber(info, params, operation) {
+		let isProvider = this.isProvidererView(info, params);
+		let ret = (isProvider && operation.isSubscribe()) || (!isProvider && !operation.isSubscribe());
+		//console.log(`isRealSubscriber: isProvider: ${isProviderer} isSubscribe: ${operation.isSubscribe()}`);
```
Comment (fmvilas) [inline lib/templateUtil.js:69]: 🚜

File: lib/templateUtil.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	getRealSubscriber(info, params, channel) {
+		let isProvider = this.isProvidererView(info, params);
+		return isProvider ? channel.subscribe() : channel.publish();
+	}
+
+}
+
+// This is the set of Java reserved words, to ensure that we don't generate reserved words.
```
Comment (fmvilas) [inline lib/templateUtil.js:85]:
```suggestion
// This is the set of Python reserved words, to ensure that we don't generate reserved words.
```

File: template/main.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
+import configparser
+import logging
+import time
+
+import messaging
+
+{# for component in asyncapi.components().s -#}
```
Comment (fmvilas) [inline template/main.py:8]: 🚜

File: lib/templateUtil.js (hunk 5 above)
Comment (damaru-inc) [reply lib/templateUtil.js]: Done

File: hooks/post-process.js (hunk 1 above)
Comment (MichaelDavisSolace) [reply hooks/post-process.js]: Was it just the commented-out code here that was the problem? (I'll remove line 9 too as info isn't being used.)

Comment (fmvilas) [reply hooks/post-process.js]: Yes, it was just about the commented code.

**Pull request merged**